README.md
# World Bank data in `R`

The `WDI` package allows users to search and download data from over 40 datasets hosted by the World Bank, including the World Development Indicators ('WDI'), International Debt Statistics, Doing Business, Human Capital Index, and Sub-national Poverty indicators.

<!-- badges: start -->
[![Travis-CI Build Status](https://travis-ci.org/vincentarelbundock/WDI.svg?branch=master)](https://travis-ci.org/vincentarelbundock/WDI)
[![AppVeyor build status](https://ci.appveyor.com/api/projects/status/github/vincentarelbundock/WDI?branch=master&svg=true)](https://ci.appveyor.com/project/vincentarelbundock/WDI)
![CRAN downloads](http://cranlogs.r-pkg.org/badges/grand-total/WDI)
<!-- badges: end -->

# Installation

`WDI` is published on CRAN and so can be installed by simply typing this in the `R` console: 

```r
install.packages('WDI')
```

To install the development version of the package, use `remotes`:

```r
library(remotes)
install_github('vincentarelbundock/WDI')
```

# Searching for data

You can search for data by using keywords in `WDIsearch`. For instance, if you are looking for data on Gross Domestic Product: 

```r
WDIsearch('gdp')
```

Which produces this: 

```r
> WDIsearch('gdp')[1:10,]
      indicator              name                                                                      
 [1,] "BG.GSR.NFSV.GD.ZS"    "Trade in services (% of GDP)"                                            
 [2,] "BM.KLT.DINV.GD.ZS"    "Foreign direct investment, net outflows (% of GDP)"                      
 [3,] "BN.CAB.XOKA.GD.ZS"    "Current account balance (% of GDP)"                                      
 [4,] "BN.CUR.GDPM.ZS"       "Current account balance excluding net official capital grants (% of GDP)"
 [5,] "BN.GSR.FCTY.CD.ZS"    "Net income (% of GDP)"                                                   
 [6,] "BN.KLT.DINV.CD.ZS"    "Foreign direct investment (% of GDP)"                                    
 [7,] "BN.KLT.PRVT.GD.ZS"    "Private capital flows, total (% of GDP)"                                 
 [8,] "BN.TRF.CURR.CD.ZS"    "Net current transfers (% of GDP)"                                        
 [9,] "BNCABFUNDCD_"         "Current Account Balance, %GDP"                                           
[10,] "BX.KLT.DINV.WD.GD.ZS" "Foreign direct investment, net inflows (% of GDP)" 
```

`WDIsearch` uses grep and ignores cases, so you can also use regular expressions. For instance, if you are looking for GDP per capita in constant dollars: 

```r
WDIsearch('gdp.*capita.*constant')
     indicator           name                                                 
[1,] "GDPPCKD"           "GDP per Capita, constant US$, millions"             
[2,] "NY.GDP.PCAP.KD"    "GDP per capita (constant 2000 US$)"                 
[3,] "NY.GDP.PCAP.KN"    "GDP per capita (constant LCU)"                      
[4,] "NY.GDP.PCAP.PP.KD" "GDP per capita, PPP (constant 2005 international $)"
```

# Download and use the data

Download a series you like for the countries you like:

```r
dat = WDI(indicator='NY.GDP.PCAP.KD', country=c('MX','CA','US'), start=1960, end=2012)
```

Look at the data: 

```r
head(dat)
  iso2c country NY.GDP.PCAP.KD year
1    CA  Canada       9374.883 1960
2    CA  Canada       9479.824 1961
3    CA  Canada       9967.366 1962
4    CA  Canada      10290.362 1963
5    CA  Canada      10774.653 1964
6    CA  Canada      11283.606 1965
```

Plot the data:

```r
library(ggplot2)
ggplot(dat, aes(year, NY.GDP.PCAP.KD, color=country)) + geom_line() + 
    xlab('Year') + ylab('GDP per capita')
```

![GDP per capita in North America](https://github.com/vincentarelbundock/WDI/raw/master/web/gdp_per_capita.jpg)

Note: You can use `country='all'` to download data for all available countries. You can also feed a vector of indicator strings if you want to download multiple indicators at once.

# Monthly or quarterly data

Some World Bank series are available at the monthly or quarterly frequency. You can download those simply using the `start` and `end` arguments:

```r
WDI(indicator = 'DPANUSSPB', country = 'CHN', start = '2012M01', end = '2012M05')

  iso2c country DPANUSSPB    year
1   CHN   China  6.324130 2012M05
2   CHN   China  6.303810 2012M04
3   CHN   China  6.313545 2012M03
4   CHN   China  6.300286 2012M02
5   CHN   China  6.313091 2012M01
```

# Automatic rename

If the vector that you supply to ``WDI`` is named, the function will automatically rename columns where possible.

```r
dat <- WDI(indicator = c("gdp_per_capita" = "NY.GDP.PCAP.KD",
                         "population" = "SP.POP.TOTL"))
head(dat)
iso2c    country year gdp_per_capita population
1    1A Arab World 2005       5378.379  316264728
2    1A Arab World 2006       5594.899  323773264
3    1A Arab World 2007       5711.663  331653797
4    1A Arab World 2008       5898.516  339825483
5    1A Arab World 2009       5782.422  348145094
6    1A Arab World 2010       5916.330  356508908
```

# Updating series list

To speed up search, `WDI` ships with a local list of all available WDI series. This list will be updated semi-regularly, but you may still want to update it manually to get access to the very latest data series. To do so, use the `cache` function:

```r
new_cache = WDIcache()
WDIsearch('gdp', cache=new_cache)
```

# Bugs, suggestions, etc.

Thanks for using `WDI`! Please send all bug reports and suggestions through the github issue tracker or by email to vincent.arel-bundock@umontreal.ca


NEWS
Version 2.7.0 (2020-06-21)
--------------------------

* API v2
* Fix a breaking change after API v2
* per_page not large enough when dataset is big
* Monthly and quarterly data

Version 2.6.0 (2019-03-02)
--------------------------

* `start` and `end` are NULL as default and download all available years.
* `label` attribute to the return data.frame includes descriptive labels for each
* `WDIbulk` function to download all of WDI
* WDIcache and data trim whitespace from region names

Version 2.5.1 (2018-12-05)
--------------------------

* New version numbering scheme
* Updated country and indicator data

Version 2.5.0 (2018-04-09)
--------------------------

* Namespace issues (thanks to Jan Dietrich)
* Updated the list of available indicators
* Allow iso3c or iso2c as input for country argument
* Rolled-back indicator input clean up because too aggressive (Thanks to Geoff Wright)
* Rename columns automatically when using a named vector

Version 2.4.0
-------------

* Call RJSONIO::fromJSON explicitly to avoid conflict with JSON package (Thanks to Richard Cotton)

Version 2.3.0
-------------

* Change default start/end date per user request

Version 2.2.0
-------------

* Bug: country names were stripped of digits

Version 2.1.0
-------------

* Fixed various issues (thanks to Matthieu Stigler for bug reports and patches)
* Faster downloads when requesting multiple countries (leveraging the US;BR;CA API syntax)

Version 2.0.0
-------------

* Now uses JSON instead of XML (Faster, easier)
* WDIcache function now allows users to update the locally stored database of available WDI series
* Fixed doc examples  
* Documentation now uses roxygen2
* Moved development to github 
* Series and country info now in a single data list. utf-8 encoding fixed    
 


test/WDI_test.R
# Warning no longer works
rm(list=ls())
library(WDI)

# Country list
co = c('IN', 'US', 'BR', 'ZW')

# Cache
cache = WDIcache()
str(cache)

# Search
WDIsearch('gdp.*capita.*constant')
WDIsearch('gdp.*capita.*constant', cache=cache)

# 1 country 1 indicator
x = WDI(country='US', indicator='NY.GDP.PCAP.KD', start=1960, end=2005)
head(x)

# 3 countries 1 indicator
x = WDI(country=c('CA', 'MX', 'US'), indicator='NY.GDP.PCAP.KD', start=1960, end=2005)
head(x)

# 1 bad country (should raise warning) 
x = WDI(country=c('BADCOUNTRY_one', 'BADCOUNTRY_two', 'CA', 'MX', 'US'), indicator='NY.GDP.PCAP.KD', start=1960, end=2005)
head(x)

# Multiple indicators 1 country
x = WDIsearch('gdp.*capita')
x = WDI(country=co, indicator=x[,1], start=2002, end=2003)
head(x)

# Multiple indicators 1 country + 2 bad indicator (should raise warning)
x = WDIsearch('gdp.*capita')
x = WDI(country=c('BADCOUNTRY', 'CA', 'MX'), indicator=c('BADIND1', 'BADIND2',  x[1:4,1]))
head(x)

# All countries, multiple indicators
x = WDIsearch('gdp.*capita')
x = WDI(country='all', indicator=c('BADIND', x[1:4,1]))
head(x)

# Bad, but allowable user input
x = WDIsearch('gdp.*capita')
x = WDI(country=c('MX', 'all'), indicator=c('BADIND', x[1:4,1]))
head(x)

# Long list of countries
x = WDI(country=WDI_data[[2]][,'iso2c'])
head(x)

# Extra information
rm(x)
x = WDI(country=co, extra=TRUE)

# iso3c and iso2c
x = WDI(country=c('US', 'CAN'))


tests/testthat.R
library(testthat)
library(WDI)

test_check("WDI")


tests/testthat/test-all.R
context('All tests in one file')

library(WDI)

# breaks on travis but works locally
test_that('WDIcache', {

    cache = WDIcache()

    expect_type(cache, 'list')
    expect_gte(length(cache$series), 15000)

    # web cache = stored cache
    old <- WDIsearch('gdp.*capita.*constant')
    new <- WDIsearch('gdp.*capita.*constant', cache=cache)
    expect_identical(old, new)

})

test_that('Bad indicator' , {
    expect_message(expect_s3_class(
        WDI(indicator = c('NY.GDP.PCAP.KD', 'test'), country = 'US', start = 1990, end = 1991),
        'data.frame'))
    expect_message(expect_null(
        WDI(indicator = c('bad', 'test'), country = 'US', start = 1990, end = 1990)))
})

test_that('Bad country', {
    expect_error(WDI(country = c('bad')))
    expect_warning(WDI(country = c('CA', 'bad'), start = 1990, end = 1991))
})

test_that('Bad year', {
    expect_error(WDI(start = 1950))
    expect_error(WDI(start = 1991, end = 1990))
    expect_error(WDI(start = '1990M2', end = '1990M1'))
})

test_that('WDI', {

    # 1 country 1 indicator
    x <- WDI(country='US', indicator='NY.GDP.PCAP.KD', start = 1990, end = 1991) 
    expect_s3_class(x, 'data.frame')
    expect_equal(dim(x), c(2, 4))

    # 3 countries 1 indicator
    x <- WDI(country=c('CA', 'MX', 'US'), 
             indicator='NY.GDP.PCAP.KD', 
             start=1990, end=1991)
    expect_s3_class(x, 'data.frame')
    expect_equal(dim(x), c(6, 4))
    expect_equal(length(unique(x$iso2c)), 3)

    # multiple indicators
    ind <- c("NY.GDP.PCAP.KD", "NY.GDP.PCAP.KN", "NY.GDP.PCAP.PP.KD")
    x <- WDI(indicator = ind, start=1990, end=1991)
    expect_gte(nrow(x), 30)
    expect_equal(ncol(x), 6)

    # bad countries
    co <- c('BADCOUNTRY_one', 'BADCOUNTRY_two', 'CA', 'MX', 'US')
    x <- expect_warning(WDI(country = co, start=1990, end=1991))
    expect_equal(length(unique(x$iso2c)), 3)

    # bad indicator 
    ind <- c("blah blah", "NY.GDP.PCAP.KD", "NY.GDP.PCAP.KN", "NY.GDP.PCAP.PP.KD", "NY.GDP.PCAP.PP.KD.87")
    x <- expect_message(WDI(indicator = ind, start=1990, end=1991))
    expect_gte(nrow(x), 30)
    expect_equal(ncol(x), 6)

})

test_that('WDI(extra = TRUE)', {

    x <- WDI(country='US', indicator='NY.GDP.PCAP.KD', 
             start=1991, end=1992, extra = TRUE)
    expect_s3_class(x, 'data.frame')
    expect_equal(dim(x), c(2, 11))

})


R/WDI_data.R
#' World Development Indicators series and country information 
#'
#' A list of two character matrices. 
#'
#' The first character matrix includes a full list of WDI series (updated 2012-06-18):
#' 
#' \itemize{
#'   \item id WDI indicator code  
#'   \item name Short name of the series
#'   \item sourceNote Detailed description provided by the source
#'   \item source Source database
#'   \item sourceOrganization Source organization
#' }
#' 
#' The second character matrix includes extra country information (updated 2012-06-18):
#'
#' \itemize{
#'   \item iso3c ISO 3-letter country code  
#'   \item iso2c ISO 2-letter country code  
#'   \item country Country name (long English form)
#'   \item region Region
#'   \item capital Capital 
#'   \item longitude Longitude 
#'   \item latitude Latitude
#'   \item income Income level as defined by the World Bank
#'   \item lending Lending category, as defined by the World Bank
#' }

#' @docType data
#' @keywords datasets
#' @name WDI_data 
#' @usage WDI_data
#' @encoding UTF-8 
#' @format List of 2 data frames 
NULL



R/WDI-package.R
#' Search, extract and format data from the World Bank's World Development Indicators
#' 
#' The \code{WDI} function provides convenient access to over 40 databases
#' hosted by the World Bank, including the World Development Indicators (WDI),
#' International Debt Statistics, Doing Business, Human Capital Index, and
#' Sub-national poverty indicators.  For fast searching, the WDI package ships
#' with a local list of available data series. This local list can be updated
#' to the latest version using the \code{WDIcache} function. 
#' 
#' @name WDI-package
#' @aliases WDI-package
#' @title Search, extract, and format data from the World Bank
#' @author Vincent Arel-Bundock \email{vincent.arel-bundock@umontreal.ca}
#' @references
#'     \url{http://arelbundock.com}
#'     \url{https://github.com/vincentarelbundock/WDI}
#' @docType package
#' @keywords package
NULL


R/WDI.R
# https://stackoverflow.com/questions/9439256/how-can-i-handle-r-cmd-check-no-visible-binding-for-global-variable-notes-when#comment20826625_12429344
# 2012 hadley says "globalVariables is a hideous hack and I will never use it"
# 2014 hadley updates his own answer with globalVariables as one of "two solutions"
globalVariables(c('year', 'value', 'Country.Name', 'Country.Code', 'Indicator.Name', 'Indicator.Code'))

#' WDI: World Development Indicators (World Bank)
#' 
#' Downloads the requested data by using the World Bank's API, parses the
#' resulting XML file, and formats it in long country-year format. 
#' 
#' @param country Vector of countries (ISO-2 character codes, e.g. "BR", "US",
#'     "CA") for which the data is needed. Using the string "all" instead of
#'     individual iso codes pulls data for every available country.
#' @param indicator Character vector of indicators codes. See the WDIsearch()
#' function. If you supply a named vector, the indicators will be automatically
#' renamed: `c('women_private_sector' = 'BI.PWK.PRVS.FE.ZS')`
#' @param start Start date, usually a year in integer format. Must be 1960 or
#' greater.
#' @param end End date, usually a year in integer format. Must be greater than
#' the `start` argument.
#' @param extra TRUE returns extra variables such as region, iso3c code, and
#'     incomeLevel
#' @param cache NULL (optional) a list created by WDIcache() to be used with the
#'     extra=TRUE argument
#' @return Data frame with country-year observations. You can extract a
#' data.frame with indicator names and descriptive labels by inspecting the
#' `label` attribute of the resulting data.frame: `attr(dat, 'label')`
#' @author Vincent Arel-Bundock \email{vincent.arel-bundock@umontreal.ca}
#' @importFrom RJSONIO fromJSON
#' @export
#'
#' @examples
#'
#' \dontrun{
#'
#' WDI(country="all", indicator=c("AG.AGR.TRAC.NO","TM.TAX.TCOM.BC.ZS"),
#'     start=1990, end=2000)
#' WDI(country=c("US","BR"), indicator="NY.GNS.ICTR.GN.ZS", start=1999, end=2000,
#'     extra=TRUE, cache=NULL)
#'
#' # Rename indicators on the fly
#' WDI(country = 'CA', indicator = c('women_private_sector' = 'BI.PWK.PRVS.FE.ZS',
#'                                   'women_public_sector' = 'BI.PWK.PUBS.FE.ZS'))
#' }
#'
WDI <- function(country = "all", 
                indicator = "NY.GDP.PCAP.KD",
                start = 1960, 
                end = 2020, 
                extra = FALSE, 
                cache = NULL){

    # Sanity: country
    if (!is.character(country)) {
        stop('The `country` argument must be a character vector')
    }

    country <- gsub('[^a-zA-Z0-9]', '', country)
    country_good <- unique(c(WDI::WDI_data$country[,'iso3c'], 
                             WDI::WDI_data$country[,'iso2c']))
    country_good <- c('all', country_good)
    country_bad <- base::setdiff(country, country_good)
    country <- base::intersect(country, country_good)

    if (length(country) == 0) {
        stop('None of the countries requested are valid. Please use ISO-2, ISO-3, or World Bank regional codes.')
    }
    if (length(country_bad) > 0) {
        warning('Please use ISO-2, ISO-3, or World Bank regional codes. Some of the country codes that you requested are invalid: ',  
                paste(country_bad, collapse = ', '))
    }

    # Sanity: start & end
    if(!(start <= end)){
        stop('`end` must be equal to or greater than `start`.')
    }
    is_integer <- function(x) is.numeric(x) && (x %% 1 == 0)
    if (is_integer(start) && (start < 1960)) {
        stop('`start` must be equal to or greater than 1960')
    }

    # Download
    dat <- list()
    failed <- NULL
    for (i in indicator) {
        tmp <- tryCatch(wdi.dl(i, country, start, end), error = function(e) e)
        if (is.null(tmp) || !inherits(tmp$data, 'data.frame') || (nrow(tmp$data) == 0)) {
            failed <- c(failed, i)
        } else {
            dat[[i]] <- tmp
        }
    }

    # Sanity: downloaded data
    if (length(dat) == 0) {
        message('None of the indicators your requested could be downloaded. Please verify the arguments of the `WDI()` function. You can also type a URL of this form in your browser to check if the World Bank web API is currently serving the indicator(s) of interest: ',
                wdi.query(indicator = failed[1])[1])
        return(NULL)
    } 

    # not downloaded and not in cache
    indicator_good <- WDI::WDI_data$series[, 'indicator']
    tmp <- base::setdiff(failed, indicator_good) 
    if (length(tmp) > 0) {
        message('These indicators could not be downloaded, and they do not appear in the cached list of available World Bank data series: ',
                paste(tmp, collapse = ', '))
    }

    # not downloaded
    tmp <- base::setdiff(failed, tmp)
    if (length(tmp) > 0) {
        message('These indicators could not be downloaded: ',
                paste(tmp, collapse = ', '))
    }

    # Extract labels
    lab = lapply(dat, function(x) data.frame('indicator' = x$indicator,
                                             'label' = x$label,
                                             stringsAsFactors = FALSE))
    lab = do.call('rbind', lab)

    # Extract data
    dat = lapply(dat, function(x) x$data)
    dat = Reduce(function(x,y) merge(x,y,all=TRUE), dat)
    row.names(dat) <- NULL

    # Extras
    if(!is.null(cache)){
        country_data = cache$country
    }else{
        country_data = WDI::WDI_data$country
    }
    if(extra==TRUE){
	    dat = merge(dat, country_data, all.x=TRUE)
    }
    countries = country[country != 'all' & !(country %in% dat$iso2c)]
    if(length(countries) > 0){
    }

    # Assign label attributes
    for (i in 1:nrow(lab)) {
        if (lab$indicator[i] %in% colnames(dat)) {
            attr(dat[[lab$indicator[i]]], 'label') = lab$label[[i]]
        }
    }

	# Rename columns based on indicator vector names
	if (!is.null(names(indicator))) {
		for (i in seq_along(indicator)) {
			idx = match(indicator[i], colnames(dat))
			if (!is.na(idx)) {
				colnames(dat)[idx] = names(indicator)[i]
			}
		}
	}

	# Output
    return(dat)
}

#' Download all the WDI indicators at once.
#' 
#' @return Data frame 
#' @author Vincent Arel-Bundock \email{vincent.arel-bundock@umontreal.ca}
#' @return a list of 6 data frames: Data, Country, Series, Country-Series,
#' Series-Time, FootNote
#' @export
WDIbulk = function() {
    if (!'tidyr' %in% utils::installed.packages()[, 1]) {
        stop('To use the `WDIbulk` function, you must install the `tidyr` package.')
    }

    # download
    temp = tempfile()
    url = 'https://databank.worldbank.org/data/download/WDI_csv.zip'
    utils::download.file(url, temp)

    # read
    zip_content = c("WDIData.csv", "WDICountry.csv", "WDISeries.csv",
                    "WDICountry-Series.csv", "WDISeries-Time.csv",
                    "WDIFootNote.csv")
    out = lapply(zip_content, function(x) utils::read.csv(unz(temp, x), stringsAsFactors = FALSE))

    # flush
    unlink(temp)

    # names
    names(out) = zip_content
    names(out) = gsub('.csv', '', names(out))
    names(out) = gsub('WDI', '', names(out))

    # clean "Data" entry
    out$Data$X = NULL
    out$Data = tidyr::gather(out$Data, year, value, -Country.Name,
                             -Country.Code, -Indicator.Name, -Indicator.Code)

    # clean year column
    out$Data$year = gsub('^X', '', out$Data$year)
    out$Data$year = as.integer(out$Data$year)

    # output
    return(out)
}

#' Internal function to build API call
#'
#' @export
#' @keywords internal
wdi.query = function(indicator = "NY.GDP.PCAP.CD", 
                     country = 'all', 
                     start = 1960, 
                     end = 2020) {

    country <- paste(country, collapse = ';')

    # WDI only allows 32500 per_page (this seems undocumented)
    out = paste0("https://api.worldbank.org/v2/country/", country, "/indicator/", indicator,
                 "?format=json",
                 "&date=", start, ":", end,
                 "&per_page=32500",
                 "&page=",1:10)
    return(out)
}

#' Internal function to download data
#'
#' @export
#' @keywords internal
wdi.dl = function(indicator, country, start, end){
    get_page <- function(daturl) {
        # download
        dat_raw = RJSONIO::fromJSON(daturl, nullValue=NA)[[2]]
        # extract data 
        dat = lapply(dat_raw, function(j) cbind(j$country[[1]], j$country[[2]], j$value, j$date))
        dat = data.frame(do.call('rbind', dat), stringsAsFactors = FALSE)
        colnames(dat) = c('iso2c', 'country', as.character(indicator), 'year')
        dat$label <- dat_raw[[1]]$indicator['value']
        # output
        return(dat)
    }

    pages <- wdi.query(indicator, country, start, end)

    dat <- list()
    done <- FALSE # done when pages no longer return useable info
    for (i in seq_along(pages)) {
        if (!done) {
            tmp <- tryCatch(get_page(pages[i]), error = function(e) NULL)
            if (inherits(tmp, 'data.frame') && (nrow(tmp) > 0)) {
                dat[[i]] <- tmp
            } else {
                done <- TRUE
            }
        }
    }
    dat <- do.call('rbind', dat)

    # numeric types
    dat[[indicator]] <- as.numeric(dat[[indicator]])

    # date is character for monthly/quarterly data, numeric otherwise
    if (!any(grepl('M|Q', dat$year))) {
        dat$year <- as.integer(dat$year)
    }

    # Bad data in WDI JSON files require me to impose this constraint
    dat = dat[!is.na(dat$year) & dat$year <= end & dat$year >= start,]

    # output
    out = list('data' = dat[, 1:4],
               'indicator' = indicator,
               'label' = dat$label[1])

    return(out)
}

#' Update the list of available WDI indicators
#'
#' Download an updated list of available WDI indicators from the World Bank website. Returns a list for use in the \code{WDIsearch} function. 
#' 
#' @return Series of indicators, sources and descriptions in two lists list  
#' @note Downloading all series information from the World Bank website can take time.
#' The \code{WDI} package ships with a local data object with information on all the series
#' available on 2012-06-18. You can update this database by retrieving a new list using \code{WDIcache}, and  then
#' feeding the resulting object to \code{WDIsearch} via the \code{cache} argument. 
#' @export
WDIcache = function(){
    # Series
    series_url = 'https://api.worldbank.org/v2/indicator?per_page=25000&format=json'
    series_dat    = RJSONIO::fromJSON(series_url, nullValue=NA)[[2]]
    series_dat = lapply(series_dat, function(k) cbind(
                        'indicator'=k$id, 'name'=k$name, 'description'=k$sourceNote, 
                        'sourceDatabase'=k$source[2], 'sourceOrganization'=k$sourceOrganization)) 
    series_dat = do.call('rbind', series_dat)          
    # Countries
    country_url = 'https://api.worldbank.org/v2/countries/all?per_page=25000&format=json'
    country_dat = RJSONIO::fromJSON(country_url, nullValue=NA)[[2]]
    country_dat = lapply(country_dat, function(k) cbind(
                         'iso3c'=k$id, 'iso2c'=k$iso2Code, 'country'=k$name, 'region'=k$region[2],
                         'capital'=k$capitalCity, 'longitude'=k$longitude, 'latitude'=k$latitude, 
                         'income'=k$incomeLevel[2], 'lending'=k$lendingType[2])) 
    country_dat = do.call('rbind', country_dat)
    row.names(country_dat) = row.names(series_dat) = NULL
    out = list('series'=series_dat, 'country'=country_dat)
    out$series = iconv(out$series, to = 'utf8')
    out$country = iconv(out$country, to = 'utf8')
    # some regions have extra whitespace in wb data
    out$country[, 'region'] = base::trimws(out$country[, 'region'])
    return(out)
}

#' Search names and descriptions of available WDI series
#' 
#' Data frame with series code, name, description, and source for the WDI series
#' which match the given criteria
#' 
#' @param string Character string. Search for this string using \code{grep} with
#'     \code{ignore.case=TRUE}.
#' @param field Character string. Search this field. Admissible fields:
#'     'indicator', 'name', 'description', 'sourceDatabase', 'sourceOrganization'
#' @param short TRUE: Returns only the indicator's code and name. FALSE: Returns
#'     the indicator's code, name, description, and source.
#' @param cache Data list generated by the \code{WDIcache} function. If omitted,
#'     \code{WDIsearch} will search a local list of series.  
#' @return Data frame with code, name, source, and description of all series which
#'     match the criteria.  
#' @export
#' @examples
#' \dontrun{
#' WDIsearch(string='gdp', field='name', cache=NULL)
#' WDIsearch(string='AG.AGR.TRAC.NO', field='indicator', cache=NULL)
#' }
WDIsearch <- function(string="gdp", field="name", short=TRUE, cache=NULL){
    if(!is.null(cache)){ 
        series = cache$series    
    }else{
        series = WDI::WDI_data$series
    }
    matches = grep(string, series[,field], ignore.case=TRUE)
    if(short){
        out = series[matches, c('indicator', 'name')]
    }else{
        out = series[matches,]
    }
    return(out)
}
